shared/src/commonMain/kotlin/com/vml/tutorial/plantshop/plants/data/FilePlantsDataSource.kt

```diff
@@ -1,17 +1,34 @@
 package com.vml.tutorial.plantshop.plants.data
 
+import com.vml.tutorial.plantshop.MR.files.plants
+import com.vml.tutorial.plantshop.core.data.FileReader
 import com.vml.tutorial.plantshop.core.data.toPlants
 import com.vml.tutorial.plantshop.plants.domain.Plant
 import com.vml.tutorial.plantshop.plants.domain.PlantsDataSource
-import com.vml.tutorial.plantshop.MR.files.plants
-import com.vml.tutorial.plantshop.core.data.FileReader
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.IO
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.launch
 
+class FilePlantsDataSource(private val fileReader: FileReader) : PlantsDataSource {
+    override fun getPlants(favoriteIds: Flow<List<Long>>): List<Plant> {
```

Comment:
Not bad :) but try to keep it simple. It could have only one responsibility -> reading or altering (not in this case) plants from this source. Setting the favorite state is another responsibility. You will / should map or filter plants with favorite state in the repository.

Reply:
Mapping is moved to the repository